### PR TITLE
BL-7562: Test reminder for an HGV/PSV

### PR DESCRIPTION
**Pull request by mrogowska**

### CODE REVIEW

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        return registration;
+    }
+
+    private Optional<uk.gov.dvsa.mot.vehicle.api.Vehicle> getMotVehicle(String registration) {
+        List<uk.gov.dvsa.mot.vehicle.api.Vehicle> motVehicles = new ArrayList<>();
+
+        try {
+            motVehicles = vehicleReadService.findByRegistration(registration);
+        } catch(Exception e) {
+            logger.error(String.format("There was an error retrieving the mot vehicle with reg: %s - MotrRequestHandler - getMotVehicle(); %s",
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: We catch the exception here and do not re-throw it. When error is raised here we would need to handle it the same way it has been handled in trade api before

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        return Optional.ofNullable(motVehicles.get(0));
+    }
+
+    private Optional<DvlaVehicle> getDvlaVehicle(String registration) {
+        DvlaVehicle dvlaVehicle = new DvlaVehicle();
+
+        try {
+            dvlaVehicle = vehicleReadService.getDvlaVehicleByRegistrationWithVin(registration);
+        } catch(Exception e) {
+            logger.error(String.format("There was an error retrieving the dvla vehicle with reg: %s - MotrRequestHandler - getDvlaVehicle(); %s",
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: same here, we will need some real error handling here

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        Vehicle foundVehicle = null;
+
+        try {
+            if(vin != null && !vin.isEmpty()) {
+                foundVehicle = hgvVehicleProvider.getVehicle(vin);
+            } else {
+                logger.error("Passed VIN was empty or null - MotrRequestHandler - getHgvPsvVehicle()");
+            }
+        } catch (Exception e) {
+            logger.error("There was an error retrieving the HGV/PSV history - MotrRequestHandler - getHgvPsvVehicle()", e);
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: same here, we will need some real error handling here

File: api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+public class HgvConfiguration {
+
+    private String apiUrl;
+    private String apiKey;
+    private int timeoutMs;
+
+    public HgvConfiguration() throws IOException {
+        this.apiKey = ConfigManager.getEnvironmentVariable(ConfigKeys.HgvPsvApiKey);
+        this.apiUrl = ConfigManager.getEnvironmentVariable(ConfigKeys.HgvPsvApiUrl);
+        this.timeoutMs = Integer.parseInt(ConfigManager.getEnvironmentVariable(ConfigKeys.HgvPsvApiConnectionTimeoutMs));
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvConfiguration.java]: TODO: verify whether we need to call HGV/PSV via proxy. If yes we will need proxy url passed in here as well

Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvConfiguration.java]: Seems like proxy is already configured under PROXY_HOST, PROXY_PORT, so try calling it the way it is used here in trade api

File: api/src/main/java/uk/gov/dvsa/mot/persist/jdbc/DvlaVehicleReadSql.java [hunk 5]
```diff
@@ -59,6 +59,38 @@
...
+            "  LEFT JOIN `dvla_make` ON `dvla_vehicle`.`make_code` = `dvla_make`.`code` " +
+            "  LEFT JOIN `dvla_model_model_detail_code_map` `map` on `map`.`dvla_make_code` = `dvla_vehicle`.`make_code` and `map`" +
+            "    .`dvla_model_code` = `dvla_vehicle`.`model_code` " +
+            "  LEFT JOIN `model` on `map`.`model_id` = `model`.`id` " +
+            "  LEFT JOIN `make` on `map`.`make_id` = `make`.`id` and `model`.`make_id` = `make`.`id` " +
+            "  JOIN `fuel_type` ON `dvla_vehicle`.`propulsion_code` = `fuel_type`.`dvla_propulsion_code` " +
+            "WHERE `dvla_vehicle`.`registration` = ? " +
+            "  AND `dvla_vehicle`.`vehicle_id` IS NULL " +
+            "ORDER BY `dvla_vehicle`.`last_updated_on` DESC " +
+            "LIMIT 1";
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/persist/jdbc/DvlaVehicleReadSql.java]: We can probably get away with removing some unused fields from here (the stuff we are getting from HGV/PSV endpoints)

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    public HgvPsvVehicle getVehicle(AwsProxyRequest request) throws Exception {
+        String registration = getRegistrationParam(request);
+
+        Optional<uk.gov.dvsa.mot.vehicle.api.Vehicle> motVehicleOptional = getMotVehicle(registration);
+        Optional<DvlaVehicle> dvlaVehicleOptional = getDvlaVehicle(registration);
+
+        if (!motVehicleOptional.isPresent() && dvlaVehicleOptional.isPresent()) {
+            DvlaVehicle dvlaVehicle = dvlaVehicleOptional.get();
+
+            Optional<Vehicle> hgvPsvVehicleOptional = getHgvPsvVehicle(dvlaVehicle.getVin());
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: If we have MOT data we shouldnt call HGV PSV

Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: nvm, did not see `!` there. But there is one logic gap I think: 

If the vehicle is a MOT vehicle then we still should respond with a proper MOT response (just like it was done before, but with additional vehicleType = "MOT" field). In this case don't think we need to call `getDvlaVehicle` anymore.

If we have no `motVehicle` data, but have `dvlaVehicle` data then we call HGV/PSV. If we get a valid HGV/PSV response then we return that, if no data was found we return `dvlaVehicle` instead.

So to sum it up: 
* We return MOT data if we have it, but if not:
* We return HGV/PSV data if we have it (we need to search dvla data first though, to find VIN), but if not:
* We return DVLA data if we have it, but if not:
* No vehicle found as it was handled previously by trade api

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+                determineMotTestNumber(vehicle)
+        );
+    }
+
+    private String determineAnnualTestExpiryDate(Vehicle vehicle) {
+        TestHistory[] testHistory = vehicle.getTestHistory();
+
+        if (testHistory.length == 0) {
+            if (vehicle.getVehicleType().equals(HGV_VEHICLE_TYPE)) {
+                return LocalDate.parse(vehicle.getRegistrationDate()).plusMonths(12).minusDays(1).toString();
```
Comment (zakrzewskam) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: This method will not generate results consistent with AC.

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            } else {
+                return LocalDate.parse(vehicle.getRegistrationDate()).plusMonths(12).toString();
```
Comment (zakrzewskam) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: This method will not generate results consistent with AC.

File: api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import uk.gov.dvsa.mot.trade.api.TradeExceptionMapper;
+
+/**
+ * Entry point for Lambda
+ */
+public class LambdaHandler implements RequestHandler<AwsProxyRequest, AwsProxyResponse> {
+
+    private static final ResourceConfig jerseyApplication = new ResourceConfig()
+            .packages("uk.gov.dvsa.mot.app")
+            .register(JacksonFeature.class)
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java]: what do we need JacksonFeature for ?

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+
+            ApiGatewayRequestContext context = (ApiGatewayRequestContext) requestContext.getProperty(
+                    RequestReader.API_GATEWAY_CONTEXT_PROPERTY);
+            if (context != null) {
+                awsRequestId = context.getRequestId();
+            }
+
+            logger.info(String.format("Entering MotrRequestHandler, awsRequestId: %s", awsRequestId));
+
+            if (Strings.isNullOrEmpty(registration)) {
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: are we sure that that method will be entered with registration as a path param will not be there ?

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+                if (!hgvPsvVehicleOptional.isPresent()) {
+                    logger.error("No HGV/PSV vehicle retrieved");
+                    throw new InvalidResourceException("No HGV/PSV vehicle retrieved", awsRequestId);
+                }
+
+                HgvPsvVehicle hgvPsvVehicle = setResponseHgvVehicle(hgvPsvVehicleOptional.get(), dvlaVehicle);
+
+                return Response.ok(hgvPsvVehicle).build();
+            } else {
+                logger.error("No MOT Test or DVLA vehicle found for registration.");
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: do we want to consider lack of vehicle as an error ?

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+                throw new InvalidResourceException("No MOT Test or DVLA vehicle found for registration " + registration,
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java:101]: InvalidResourceException - what is the effect of throwing that exception ? What response will be produced ?

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+                        awsRequestId);
+            }
+        } catch (TradeException e) {
+            // no need to log these errors, just throw them back
+            logger.trace("Exiting getVehicle");
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: finally clause looks to be best home for that log statement

File: api/src/main/java/uk/gov/dvsa/mot/trade/read/core/MotrReadServiceDatabase.java [hunk 14]
```diff
@@ -35,11 +35,22 @@ public MotrReadServiceDatabase(MotTestReadService motTestReadService, VehicleRea
...
+
+    @Override
+    @ProvideDbConnection
+    public MotrResponse getLatestMotTestByRegistration(String registration, Boolean withDvlaVehicles) {
+
         List<Vehicle> vehicles = vehicleReadService.findByRegistration(registration);
         MotrReadServiceDatabase.VehicleAndLatestMot vehicleAndLatestMot = getVehicleAndLatestMotTestPass(vehicles);
 
         if (vehicleAndLatestMot == null || !vehicleAndLatestMot.hasMotTest()) {
-            return getDvlaVehicleByRegistration(registration);
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/trade/read/core/MotrReadServiceDatabase.java]: let's split that one. DvlaVehicle does not have to have MotTest at all so logically it is hard to merge those two. Boolean flags as parameters to the method - noooo :-)

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+            Optional<MotrResponse> motVehiclesOptional = getMotVehicle(registration);
+
+            if (motVehiclesOptional.isPresent()) {
+
+                return Response.ok(setResponseMotVehicle(motVehiclesOptional.get())).build();
+            }
+
+            Optional<DvlaVehicle> dvlaVehicleOptional = getDvlaVehicle(registration);
+
+            if (dvlaVehicleOptional.isPresent()) {
```
Comment (colla) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: Untested DVLA vehicles which are not HGVs, will not be returned from any HGV API.

Does this code expect otherwise?
Will it throw an exception for an untested car?

File: api/src/main/java/uk/gov/dvsa/mot/trade/api/TradeExceptionMapper.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+@Priority(1)
+@Provider
+public class TradeExceptionMapper implements ExceptionMapper<TradeException> {
+
+    @Override
+    public Response toResponse(TradeException e) {
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/trade/api/TradeExceptionMapper.java:14]: just to clarify, when returning non 200 codes, do we return text message as response or something else like json in other parts of the code?

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+    private HgvPsvVehicle setResponseHgvVehicle(Vehicle vehicle, DvlaVehicle dvlaVehicle) throws Exception {
+        HgvPsvVehicle hgvPsvVehicle = new HgvPsvVehicle();
+
+        hgvPsvVehicle.setMake(vehicle.getMake());
+        hgvPsvVehicle.setModel(vehicle.getModel());
+        hgvPsvVehicle.setPrimaryColour(dvlaVehicle.getColour1().getName());
+        hgvPsvVehicle.setRegistration(dvlaVehicle.getRegistration());
+        hgvPsvVehicle.setVin(dvlaVehicle.getVin());
+        hgvPsvVehicle.setVehicleType(vehicle.getVehicleType());
+        hgvPsvVehicle.setManufactureYear(vehicle.getYearOfManufacture().toString());
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: make sure the fields you're chaining methods on are non-nullable, otherwise, it will fail as some data in the table can be null.

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+        if (vehicle.getTestHistory() != null && vehicle.getTestHistory().length > 0) {
+            TestHistory[] testHistory = vehicle.getTestHistory();
+            hgvPsvVehicle.setMotTestNumber(testHistory[testHistory.length - 1].getTestCertificateSerialNo());
+        }
+
+        return hgvPsvVehicle;
+    }
+
+    private String determineAnnualTestExpiryDate(Vehicle vehicle) throws Exception {
+        DateTimeFormatter oldPattern = DateTimeFormatter.ofPattern("dd/MM/yyyy");
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: that can be extracted to a separate class and tested in isolation

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+                datetime = datetime.plusMonths(12);
+            }
+        } else {
+            datetime = LocalDate.parse(date, oldPattern);
+        }
+
+        return datetime.format(newPattern);
+    }
+
+    private HgvPsvVehicle setResponseMotVehicle(MotrResponse response) {
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: we are not setting anything here. Maybe we "buildHgvPsvResponse"

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    private HgvPsvVehicle setResponseHgvVehicle(Vehicle vehicle, DvlaVehicle dvlaVehicle) throws Exception {
+        HgvPsvVehicle hgvPsvVehicle = new HgvPsvVehicle();
+
+        hgvPsvVehicle.setMake(vehicle.getMake());
+        hgvPsvVehicle.setModel(vehicle.getModel());
+        hgvPsvVehicle.setPrimaryColour(dvlaVehicle.getColour1().getName());
+        hgvPsvVehicle.setRegistration(dvlaVehicle.getRegistration());
+        hgvPsvVehicle.setVin(dvlaVehicle.getVin());
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: We don't use vin anymore, this should be removed (in other places where we added this as well)

File: api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+            vehicle = vehicleFuture.get();
+            vehicleTestHistory = vehicleTestHistoryFuture.get();
+        } catch (ExecutionException | InterruptedException e) {
+            logger.error("HGV calls execution error", e);
+            throw new Exception("HGV calls execution error" + e.getStackTrace());
+        }
+
+        if (vehicle != null) {
+            vehicle.setTestHistory(vehicleTestHistory);
+            logger.info("HGV vehicle fetched successfully. Registration " + vehicle.getVehicleIdentifier());
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java]: trace/debug

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+        return Optional.ofNullable(foundVehicle);
+    }
+
+    private HgvPsvVehicle setResponseHgvVehicle(Vehicle vehicle, DvlaVehicle dvlaVehicle) throws Exception {
+        HgvPsvVehicle hgvPsvVehicle = new HgvPsvVehicle();
+
+        hgvPsvVehicle.setMake(vehicle.getMake());
+        hgvPsvVehicle.setModel(vehicle.getModel());
+        hgvPsvVehicle.setPrimaryColour(dvlaVehicle.getColour1().getName());
+        hgvPsvVehicle.setRegistration(dvlaVehicle.getRegistration());
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: I think we should use the registration provided by motr/returned by HGV/PSV rather than the one from dvlaVehicle (like its done in MOTH)

File: api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+        TestHistory[] vehicleTestHistory;
+
+        try {
+            CompletableFuture<Vehicle> vehicleFuture = CompletableFuture.supplyAsync(() -> {
+                Vehicle result = null;
+
+                try {
+                    result = getHgvVehicle(registration);
+                } catch (IOException e) {
+                    logger.error("IO error during communication with HGV vehicle api",  e);
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java]: catch cascade is almost duplicated between vehicle and history... can we DRY it ?

File: api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+
+    @Inject
+    public void setHgvConfiguration(HgvConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    public Vehicle getVehicle(String registration) throws Exception {
+        logger.trace("Entering HgvVehicleProvider getVehicle()");
+
+        if (registration == null || registration.isEmpty()) {
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java]: I think we can guarantee registration is not empty much earlier and clean this out.

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 25]
```diff
@@ -0,0 +1,252 @@
...
+    private String determineAnnualTestExpiryDate(Vehicle vehicle) throws Exception {
+        DateTimeFormatter oldPattern = DateTimeFormatter.ofPattern("dd/MM/yyyy");
+        DateTimeFormatter newPattern = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+
+        LocalDate datetime;
+
+        String date = vehicle.getTestCertificateExpiryDate();
+
+        if (Strings.isNullOrEmpty(date)) {
+            logger.info("Test Certificate Expiry date is null or empty");
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: trace/debug

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 26]
```diff
@@ -0,0 +1,252 @@
...
+        } catch (Exception e) {
+            logger.error(String.format("There was an error retrieving mot vehicle with reg: %s",
+                    registration), e);
+            throw new InvalidResourceException("There was an error retrieving mot vehicle", awsRequestId);
+        }
+
+        return Optional.ofNullable(motVehicles);
+    }
+
+
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: too much whitespace

File: api/src/main/java/uk/gov/dvsa/mot/persist/VehicleReadDao.java [hunk 27]
```diff
@@ -43,6 +43,8 @@
 
     DvlaVehicle getDvlaVehicleByRegistration(String registration);
 
+    DvlaVehicle getDvlaVehicleByRegistrationWithVin(String registration);
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/persist/VehicleReadDao.java]: we dnot need vin anymore

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 28]
```diff
@@ -0,0 +1,252 @@
...
+        return Optional.ofNullable(foundVehicle);
+    }
+
+    private HgvPsvVehicle setResponseHgvVehicle(Vehicle vehicle, DvlaVehicle dvlaVehicle) throws Exception {
+        HgvPsvVehicle hgvPsvVehicle = new HgvPsvVehicle();
+
+        hgvPsvVehicle.setMake(vehicle.getMake());
+        hgvPsvVehicle.setModel(vehicle.getModel());
+        hgvPsvVehicle.setPrimaryColour(dvlaVehicle.getColour1().getName());
```
Comment (ryszardm-kainos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: dvlaVehicle might be different then the hgvVehicle, we should not map anything from MOT db to hgv response (we won't be displaying colour on FE for HGV/PSV)

File: api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+            .register(TradeExceptionMapper.class);
+
+    private static final JerseyLambdaContainerHandler<AwsProxyRequest, AwsProxyResponse> handler =
+            JerseyLambdaContainerHandler.getAwsProxyHandler(jerseyApplication);
+
+
+    @Override
+    public AwsProxyResponse handleRequest(AwsProxyRequest request, Context context) {
+        context.getLogger().log("Entering lambda handler : " + context.getAwsRequestId());
+
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java:25]: too much of white chars

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 30]
```diff
@@ -0,0 +1,199 @@
...
+
+@Path("/")
+public class MotrRequestHandler extends AbstractRequestHandler {
+    private static final Logger logger = Logger.getLogger(MotrRequestHandler.class);
+
+    private MotrReadService motrReadService;
+    private HgvVehicleProvider hgvVehicleProvider;
+    private String awsRequestId;
+
+    public MotrRequestHandler() {
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: superflous

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 31]
```diff
@@ -0,0 +1,199 @@
...
+        super();
+    }
+
+    public MotrRequestHandler(boolean injectSelf) {
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java:42]: is it useful at all ? And if not, empty constructor can go, too.

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 32]
```diff
@@ -0,0 +1,199 @@
...
+                Optional<Vehicle> hgvPsvVehicleOptional = getHgvPsvVehicle(registration);
+
+                if (!hgvPsvVehicleOptional.isPresent()) {
+                    logger.info("No HGV/PSV vehicle retrieved");
+
+                    if (dvlaVehicle.getMotTestExpiryDate() != null) {
+                        return Response.ok(buildMotResponse(dvlaVehicle)).build();
+                    }
+
+                    throw new InvalidResourceException("No HGV/PSV vehicle retrieved for registration " + registration,
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: what is InvalidResourceException ? Does it end up with 404 ?

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 33]
```diff
@@ -0,0 +1,199 @@
...
+                HgvPsvVehicle hgvPsvVehicle = buildHgvPsvResponse(hgvPsvVehicleOptional.get(), dvlaVehicle);
+
+                return Response.ok(hgvPsvVehicle).build();
+            } else {
+                logger.error("No MOT Test or DVLA vehicle found for registration.");
+                throw new InvalidResourceException("No MOT Test or DVLA vehicle found for registration " + registration,
+                        awsRequestId);
+            }
+        } catch (TradeException e) {
+            // no need to log these errors, just throw them back
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java:105]: what will happen with this exception when rethrown ?

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java [hunk 34]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    private Optional<Vehicle> getHgvPsvVehicle(String registration) throws Exception {
+        Vehicle foundVehicle = null;
+
+        try {
+            if (!Strings.isNullOrEmpty(registration)) {
+                foundVehicle = hgvVehicleProvider.getVehicle(registration);
+            } else {
+                logger.warn("Passed registration was empty or null");
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java:155]: is it possible to pass an empty registration ?

File: api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java [hunk 35]
```diff
@@ -0,0 +1,120 @@
...
+                    return response.readEntity(type);
+                case 204:
+                    response.close();
+                    return type.newInstance();
+                default:
+                    response.close();
+                    throw new IOException("Invalid http response code");
+            }
+        } catch (IOException e) {
+            logger.error(String.format("IO error during communication with api from HGV endpoint : %s", endpointToCall), e);
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java]: that can be collapsed into one catch clause with pipe operator. One does not need to have exception name in the message while exception is logged anyway

File: api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java [hunk 36]
```diff
@@ -0,0 +1,120 @@
...
+        } catch (InstantiationException e) {
+            logger.error(String.format("Instantiation error during communication with api from HGV endpoint : %s", endpointToCall), e);
+        } catch (IllegalAccessException e) {
+            logger.error(String.format("IllegalAccess error during communication with api from HGV endpoint : %s", endpointToCall), e);
+        }
+
+        return null;
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java]: so when 204 response code is returned, which means not found, we return type.newInstance(). How is that line different to 'not found' case ?

File: api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+import com.amazonaws.serverless.proxy.model.AwsProxyResponse;
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+
+import org.glassfish.jersey.server.ResourceConfig;
+
+import uk.gov.dvsa.mot.trade.api.TradeExceptionMapper;
+
+/**
+ * Entry point for Lambda
```
Comment (przemos) [inline api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java]: that comment has little value

File: api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java (hunk 29 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java:25]: fixed

File: api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java (hunk 37 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/app/LambdaHandler.java]: removed

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java (hunk 31 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java:42]: From what I see its a workaround for transitive Guice injections for unit tests purposes. I'm guessing the empty constructor is for guice in the live code - can check that later

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java (hunk 32 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java]: Aye, its 404 with error type "Invalid Resource".

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java (hunk 12 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java:101]: Http: 404 not found with for example:

{ "httpStatus": "404", "errorMessage": "No MOT Test or DVLA vehicle found for registration EXTODAYY", "awsRequestId": "xxx"}

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java (hunk 33 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java:105]: TradeException constructors ensure it has a status code specified. From what I can see some exceptions inherit it, for example:
InvalidResourceException - TradeException with HTTP:404
BadRequestException - TradeException with HTTP:400

Based on what kind of TradeException is caught here we will have different error responses

File: api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java (hunk 34 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/app/MotrRequestHandler.java:155]: not possible from the integrated motr->trade perspective. Might become more possible with trailers when calling api with empty string directly? Currently this would be useful only in the off-chance that there is a dvla_vehicle with registration field empty, which is possible (from DB schema perspective). 

I'd leave that in since its only two lines and has some theoretical value

File: api/src/main/java/uk/gov/dvsa/mot/trade/api/TradeExceptionMapper.java (hunk 16 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/trade/api/TradeExceptionMapper.java:14]: From what I can see we always craft a json here. Not sure why its returned as plain text

File: api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java (hunk 35 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java]: fixed

File: api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java (hunk 36 above)
Comment (ryszardm-kainos) [reply api/src/main/java/uk/gov/dvsa/mot/vehicle/hgv/HgvVehicleProvider.java]: This does not seem right. In this case its treated as if nothing was returned from HGV api. And AC states we would like an error... Will add an exception here

**Review by przemos — APPROVED** [review]
